server/src/main/java/org/elasticsearch/rest/ChunkedZipResponse.java

```diff
@@ -0,0 +1,382 @@
+     * listener.
+     *
+     * @param entryName The name of the entry in the response zip file.
+     * @param listener  A listener which is completed when the entry has been completely processed: either fully sent, or else the request
+     *                  was cancelled and the response will not be used any further. If the returned entry listener is completed
+     *                  exceptionally then the exception is passed to {@code listener}, otherwise this listener is completed successfully.
+     */
+    public ActionListener<ChunkedRestResponseBodyPart> newEntryListener(String entryName, ActionListener<Void> listener) {
+        if (listenersRefs.tryIncRef()) {
+            final var zipEntry = new ZipEntry(filename + "/" + entryName);
+            return ActionListener.assertOnce(ActionListener.releaseAfter(listener.delegateFailureAndWrap((l, firstBodyPart) -> {
+                if (firstBodyPart == null) {
+                    l.onResponse(null);
+                } else {
+                    enqueueEntry(zipEntry, firstBodyPart, l);
+                }
+            }), listenersRefs::decRef));
+        } else {
+            assert false : "already closed";
+            throw new AlreadyClosedException("response already closed");
+        }
+    }
+
+    /**
+     * Queue of entries that are ready for transmission.
+     */
+    private final Queue<ChunkedZipEntry> entryQueue = new LinkedBlockingQueue<>();
+
+    /**
+     * Upper bound on the number of entries in the queue, atomically modified to ensure there's only one thread processing queue entries
+     * at once.
+     */
+    private final AtomicInteger queueLength = new AtomicInteger();
+
+    /**
+     * Ref-counting for access to the queue, to avoid clearing the queue on abort concurrently with an entry being sent.
+     */
+    private final RefCounted queueRefs = AbstractRefCounted.of(this::drainQueue);
+
+    /**
+     * Flag to indicate if the request has been aborted, at which point we should stop enqueueing more entries and promptly clean up the
+     * ones being sent.
+     */
+    private final AtomicBoolean isRestResponseFinished = new AtomicBoolean();
+
+    private boolean tryAcquireQueueRef() {
+        return isRestResponseFinished.get() == false && queueRefs.tryIncRef();
+    }
+
+    private void enqueueEntry(ZipEntry zipEntry, ChunkedRestResponseBodyPart firstBodyPart, ActionListener<Void> listener) {
+        if (tryAcquireQueueRef()) {
+            try {
+                entryQueue.add(new ChunkedZipEntry(zipEntry, firstBodyPart, () -> listener.onResponse(null)));
+                if (queueLength.getAndIncrement() == 0) {
+                    final var nextEntry = entryQueue.poll();
+                    assert nextEntry != null;
+                    final var continuation = new QueueConsumer(nextEntry.zipEntry(), nextEntry.firstBodyPart());
```

Comment:
We have contiuations for both entries and parts. It would be great if we can make it clear in the variable name which continuation it is for?

Reply:
Ok I've tried to do that.